### PR TITLE
Change Vector to AbstractVector and add test for @view

### LINKED ISSUE #131: Can't evaluate problem functions with views
```julia
julia> nlp = ADNLPModel(x -> dot(x, x), zeros(2))
julia> x = ones(3);
julia> obj(nlp, @view x[1:2])
ERROR: MethodError: no method matching obj(::NLPModels.ADNLPModel, ::SubArray{Float64,1,Array{Float64,1},Tuple{UnitRange{Int64}},true})
```

This might also apply to other problem types.

### PR DESCRIPTION
**Pull request by abelsiqueira**
Closes #131

### CODE REVIEW

File: src/autodiff_nlsmodel.jl [hunk 1]
```diff
@@ -152,19 +152,19 @@ function hess(nls :: ADNLSModel, x :: Vector; obj_weight = 1.0, y :: Vector = []
...
-function hess_coord(nls :: ADNLSModel, x :: Vector; obj_weight = 1.0, y :: Vector = [])
+function hess_coord(nls :: ADNLSModel, x :: AbstractVector; obj_weight = 1.0, y :: AbstractVector = [])
   H = hess(nls, x, obj_weight=obj_weight, y=y)
   return findnz(H)
 end
 
-function hprod(nls :: ADNLSModel, x :: Vector, v :: Vector;
-    obj_weight = 1.0, y :: Vector = [])
+function hprod(nls :: ADNLSModel, x :: AbstractVector, v :: AbstractVector;
+    obj_weight = 1.0, y :: AbstractVector = [])
```
Comment (dpo) [inline src/autodiff_nlsmodel.jl]: Should the default value be `Float64[]`?

File: src/autodiff_nlsmodel.jl [hunk 2]
```diff
@@ -152,19 +152,19 @@ function hess(nls :: ADNLSModel, x :: Vector; obj_weight = 1.0, y :: Vector = []
...
   Hv = zeros(nls.meta.nvar)
   return hprod!(nls, x, v, Hv, obj_weight=obj_weight, y=y)
 end
 
-function hprod!(nls :: ADNLSModel, x :: Vector, v :: Vector, Hv :: Vector;
-    obj_weight = 1.0, y :: Vector = [])
+function hprod!(nls :: ADNLSModel, x :: AbstractVector, v :: AbstractVector, Hv :: AbstractVector;
+    obj_weight = 1.0, y :: AbstractVector = [])
```
Comment (dpo) [inline src/autodiff_nlsmodel.jl]: Same here?

File: src/lls_model.jl [hunk 3]
```diff
@@ -82,53 +82,53 @@ function hprod_residual!(nls :: LLSModel, x :: AbstractVector, i :: Int, v :: Ab
...
 
-function jtprod!(nls :: LLSModel, x :: Vector, v :: Vector, Jtv :: Vector)
+function jtprod!(nls :: LLSModel, x :: AbstractVector, v :: AbstractVector, Jtv :: AbstractVector)
   increment!(nls, :neval_jtprod)
   Jtv[1:nls.meta.nvar] = nls.C' * v
   return Jtv
 end
 
-function hess(nls :: LLSModel, x :: Vector; obj_weight = 1.0, y :: Vector = [])
+function hess(nls :: LLSModel, x :: AbstractVector; obj_weight = 1.0, y :: AbstractVector = [])
```
Comment (dpo) [inline src/lls_model.jl]: Same here, an several other places.

**Comment by dpo** [comment]
Many thanks for this!!! I notice that sometimes, we use `AbstractVector` and sometimes `AbstractVector{Float64}`. Is there a reason?

**Comment by abelsiqueira** [comment]
I don't remember if we discussed something about it. I noticed that `NLPModelMeta` uses `Array{Float64,1}`, and I think we used `Vector{Float64}` at first because of it, but eventually added functions with `Vector`.

I think changing `NLPModelMeta` to `AbstractVector` is in order?

**Comment by dpo** [comment]
I think it's inefficient to have abstract types inside a struct. But I think we should use either AbstractVector or AbstractVector{Float64} everywhere.

**Comment by abelsiqueira** [comment]
To handle different precision, we can't use `Float64` and to handle views, we need abstract. So I think it needs to be `AbstractVector`. Unless we don't allow views inside `NLPModelMeta`.
PS. Why is it ineficient?

**Comment by dpo** [comment]
Ok for `AbstractVector`. Thanks!

https://docs.julialang.org/en/v0.7.0/manual/performance-tips/#Avoid-containers-with-abstract-type-parameters-1

**Comment by dpo** [comment]
Another thought. Eventually, we'll want to parameterize `NLPModelMeta` and `NLPModel` to indicate what precision is used. I suggest we leave `Vector` in `NLPModelMeta` for now. Those don't need to be views or subarrays.

**Comment by abelsiqueira** [comment]
Done.

**Comment by dpo** [comment]
Thank you! Let's go with the last version of NLPModels compatible with 0.6.

**Pull request merged**